gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java

```diff
@@ -336,8 +345,14 @@ private void getNames(Node node, ImmutableSet.Builder<QualifiedName> tableBuilde
             case SetSchemaAuthorization s -> setCatalogAndSchemaNameFromSchemaQualifiedName(Optional.of(s.getSource()), catalogBuilder, schemaBuilder, catalogSchemaBuilder);
             case SetTableAuthorization s -> tableBuilder.add(qualifyName(s.getSource()));
             case SetViewAuthorization s -> tableBuilder.add(qualifyName(s.getSource()));
-            case Table s -> tableBuilder.add(qualifyName(s.getName()));
+            case Table s -> {
+                // ignore temporary tables as they can have various table parts
+                if (!temporaryTables.contains(s.getName())) {
+                    tableBuilder.add(qualifyName(s.getName()));
+                }
+            }
             case TableFunctionInvocation s -> tableBuilder.add(qualifyName(s.getName()));
+            case WithQuery withQuery -> temporaryTables.add(QualifiedName.of(withQuery.getName().getValue()));
```

Comment:
If you revert this to handle `Query` nodes, then we can bypass my concern about encountering a `Table` with a `WithQuery` `name` before it is added to `temporaryTables`. The concern is basically that `node.getChildren()` could return a list with the `QueryBody` before the `With`. If we explicitly process any `With`, then we can ensure that the names are collected before they are referenced. Something like
```
case Query query -> query.getWith().ifPresent(
                    with -> temporaryTables.addAll(with.getQueries().stream().map(WithQuery::getName).map(Identifier::getValue).map(QualifiedName::of).toList()));
```